src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -41,19 +41,54 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
                                                   std::vector<std::shared_ptr<AbstractLQPNode>>& push_down_nodes,
                                                   AbstractCardinalityEstimator& estimator) {
   const auto input_node = current_node->input(input_side);
-  if (!input_node) return;  // Allow calling without checks
+  if (!input_node) {
+    Assert(push_down_nodes.empty(), "Expected pushdown nodes to be already inserted.");
+    return;  // Allow calling without checks
+  }
 
-  // A helper method for cases where the input_node does not allow us to proceed
+  // A helper method for cases where the input_node does not allow us to proceed.
   const auto handle_barrier = [&]() {
     _insert_nodes(current_node, input_side, push_down_nodes);
 
-    if (input_node->left_input()) {
+    const auto& barrier_node = input_node;
+    const auto barrier_node_is_pushdown_predicate = [&barrier_node]() {
+      if (barrier_node->type == LQPNodeType::Predicate) {
+        const auto predicate_node = std::static_pointer_cast<PredicateNode>(barrier_node);
+        return !_is_expensive_predicate(predicate_node->predicate());
```

Comment:
Is there any case where `handle_barrier()` is called with `input_node`/`barrier_node` being a `PredicateNode`? Wouldn't it be sufficient to perform the below join mode checks only? Or in other words: when does a `PredicateNode` (not a JoinNode treated as if it is a predicate) has two inputs?